admission_control/admission-control-service/src/upstream_proxy.rs

```diff
@@ -20,184 +21,180 @@ use network::validator_network::{
 use prost_ext::MessageExt;
 use rand::seq::SliceRandom;
 use std::collections::HashMap;
+use tokio::runtime::TaskExecutor;
 
 /// Full nodes use UpstreamProxy to send transaction write requests to their upstream validator,
 /// and is essentially the networking layer of AC.
 /// UpstreamProxy is instantiated in AC Runtime to communicate with the gRPC layer.
 /// The requests and responses are represented with AdmissionControlNetworkSender and AdmissionControlNetworkEvents.
-pub(crate) struct UpstreamProxy {
+#[derive(Clone)]
+pub struct UpstreamProxyData {
     ac_config: AdmissionControlConfig,
     network_sender: AdmissionControlNetworkSender,
     peer_info: HashMap<PeerId, bool>,
-    /// used to process client requests
-    client_events: mpsc::UnboundedReceiver<(
-        SubmitTransactionRequest,
-        oneshot::Sender<failure::Result<SubmitTransactionResponse>>,
-    )>,
     role: RoleType,
     /// AC Client
     pub client: AdmissionControlClient, // TODO remove client
 }
 
-impl UpstreamProxy {
-    /// bootstrap of UpstreamProxy
+impl UpstreamProxyData {
     pub fn new(
-        config: &NodeConfig,
+        ac_config: AdmissionControlConfig,
         network_sender: AdmissionControlNetworkSender,
-        client_events: mpsc::UnboundedReceiver<(
-            SubmitTransactionRequest,
-            oneshot::Sender<failure::Result<SubmitTransactionResponse>>,
-        )>,
-        client: AdmissionControlClient,
+        peer_info: HashMap<PeerId, bool>,
```

Comment:
why do you need to clone peer_info?
can it be just owned by main routine `process_network_messages`

Reply:
Because we need to pass in peer_info into process_submit_transaction_request and submit_transaction_upstream anyway with cloning (because new and lost peer need to borrow them mutably)

Reply:
better keep peer_info as local variable in `process_network_messages`. You'll be able to pass it to lost_peer and new_peer handlers. Cause it's the same thread
As for `start_submit_transaction_upstream` you can calculate peer_id in advance in main thread and pass it to that routine without need to clone HashMap

Reply:
sg!